### PR TITLE
Paper: Widgets and Astropy: Accomplishing Productive Research with Undergraduates (Craig)

### PR DESCRIPTION
**Pull request by mwcraig**
I apologize in advance for the incomplete state of this draft...more is coming over the next couple of days.

[Edit by jaberg] http://nipy.bic.berkeley.edu:5000/download/9

### CODE REVIEW

**Comment by jeanconn** [comment]
Hi Matt,  

I've been assigned/suggested as a reviewer for your paper.  With the understanding that it is still in early draft state, I've got a few comments that may or may not be useful.  

I might suggest reorganizing your introduction to go something more like "The Students", "The Goal", "The Problem", "The Solution".  Right now, you are bouncing back and forth a bit between these ideas (Astropy, some problems, the students, some problems, IPython, the widget) and I think it would be stronger with a more streamlined flow from issues to solution.  For example, right now you are making some broad generalizations and not really supporting them "In practice, despite extensive documentation, the barrier to entry is high", and I think you could skip that if you talked a bit more about where you want the students to go (doing actual astronomy in a short period of time), the problems with the available tools (some tools don't help with reproducible astronomy, some don't work on your platforms, Astropy is there, but selection, configuring, and chaining the tools takes a lot of time), and  your solution (setting up a workflow as a widget).

You might also want to mention some of the reasons why you've gone with the widget idea (what the widget brings to the table that you wouldn't get with just some functions and setup saved in the notebook).

Oh, and capitalize Python everywhere.

Regards,
Jean

**Comment by mwcraig** [comment]
Thanks, @jeanconn, the comments so far are helpful. Will get an update done later today or first thing on Saturday. This is my first presentation at SciPy, so your suggestions are very useful.

**Comment by kbarbary** [comment]
Hi Matt! I have also been suggested as a reviewer. The draft looks good so far; I think the screenshots of the widgets will go a long way once you add them.

My one suggestion at this point is that you move the section "The reducer widget structure" to later in the paper, possibly after the "image browser" section. It's better to explain what `reducer` does before explaining how it does it.

**Comment by jakevdp** [comment]
Hi,
I think this is coming together well. I'm excited to see the screenshots added to the paper; I think that will help the reader get a better picture of how the package is used.

I also agree with the above comments that the overall organization should be re-worked as the content is filled-in. Looking forward to reviewing the final draft!

**Comment by mwcraig** [comment]
Thanks for the additional feedback Jake and Kyle -- I'll ping everyone when the draft is finalized, ETA is end of day, June 24.

**Comment by mwcraig** [comment]
Thanks for your patience, everyone. While I'm sure there are improvements to be made, this is now ready for a more thorough review.

**Comment by jeanconn** [comment]
Hi Matt,

The figures are great!  They really make the paper shine and illustrate how the widget helps you to do science with your students.

I think there is a tiny problem  in that you've run out of visual aids by the time you get to your widget description.  @kbarbary , do you think maybe he would be better off reversing the order (again) and talking about the widget structure before doing the reveal and showing off the widget?  I'm really not sure.  It could obviously work either way.

File: papers/matthew_craig/matthew_craig.rst [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+
+--------------------------------------------------------------------------
+Widgets and Astropy: Accomplishing Productive Research with Undergraduates
+--------------------------------------------------------------------------
+
+.. class:: abstract
+
+    A set of IPython notebooks with a widget interface that are based on
+    Astropy, a community-developed package of fundamental tools for astronomy,
+    and Astropy affiliated packages. The widget interface makes astropy a much
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: Might want a full sentence to start the abstract.

File: papers/matthew_craig/matthew_craig.rst [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+project requires calibrating and taking measurements from a couple of hundred
+images. To the extent possible, analysis needs to be automated. Roughly half
+of the students use Windows, the rest Mac OSX.
+
+The problem, described in more detail below, is that the GUI-based software
+most accessible to these students is expensive, often available only on
+Windows, and not clearly documented. The free options largely require
+programming. The software must also leave a record of the choices made in
+calibrating the images so that future researchers can use the images with
+confidence.
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: Should probably say the available software *doesn't* leave a record, which is required to do good science.

File: papers/matthew_craig/matthew_craig.rst [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+
+Extraction of the brightness of individual stars is called photometry. There
+are several techniques for performing photometry, all of which estimate and
+eliminate the sky background.
+
+The problem
+-----------
+
+Several software packages can calibrate astronomical images and perform
+photometry, which begs the question "Why write another one?"
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: You can, of course, do this if you want, but I prefer the traditional use of "begs the question" which means you've made a conclusion based on a premise that lacks support.  "Raises" or brings "brings up" would both be great substitutes here.  Or just ask the question yourself "photometry, so why write another one?".

File: papers/matthew_craig/matthew_craig.rst [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+called ``reduction.ipynb``, in the directory in which it is invoked.
+
+The intent is that the raw, uncalibrated images are stored in a different
+directory and that the calibrated images are in the same directory as the
+notebook, leaving a *human-readable* record with the images describing the
+choices made in calibration. The notebook does not provide an easy way to re-
+do the calibration in a particular directory. In discussions with students
+while developing ``reducer`` it became clear that it would be impossible to
+ensure that the record in the notebook matched the files in the directory
+unless it was difficult to re-run the notebook.
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: I actually don't understand what you are trying to say here, and it sounds like it is key to understanding the paper.  Did you want to make it difficult to re-run the notebook?

File: papers/matthew_craig/matthew_craig.rst [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+The back end of ``reducer`` is built on the Astropy project [Astropy2013]_, a
+community-driven effort to develop high-quality, open source tools for Python
+in astronomy, and on Astropy affiliated projects. [#]_ Astropy was chosen because it has a large developer community of professional astronomers.
+
+Section W provides background on the science of image calibration. In section
+X the problem is discussed more completely, including a review of some of the
+available options for astronomical image processing. The section Y discusses
+the use of ``reducer``, while section Z presents its implementation. The
+widget classes in ``reducer`` are potentially useful in other applications.
+
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: Do these section letters all map to names at this point?

File: papers/matthew_craig/matthew_craig.rst [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+``apply_to`` argument to the initializer controls which calibrated images the
+widget will act on; in this case its value is ``{'imagetyp': 'flat'}``, which
+selects the calibration images used to correct non-uniform illumination. A
+``group_by`` argument to the widget initializer controls controls how the
+images selected by ``apply_to`` are combined. In the example shown, all images
+with the same filter and exposure time will be combined by averaging, after
+each image has been scaled to the same median value.
+
+Each image, including the images used in the calibration itself, is processed
+by a ``ReductionWdiget``, like that shown in Fig. :ref:`light-settings`. That
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: Typo, ReductionWdiget should probably be ReductionWidget.

File: papers/matthew_craig/matthew_craig.rst [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+  to a computer.
++ A DC offset to prevent negative pixel values.
+
+The first stage of calibration is to remove the noise and offset from each
+image. The second stage is to correct for imperfections in the optical system
+that affect how much light gets to each pixel in the camera. An example of
+this sort of imperfection is dust on the camera itself.
+
+A series of images is taken and then combined to perform each type of
+calibration. *Bias* images corrects for the DC offset, *dark* images correct
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: images corrects -> images correct

File: papers/matthew_craig/matthew_craig.rst [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+               *SciPy: Open source scientific tools for Python*,
+               http://scipy.org/ 2001
+
+.. [Pérez2007] Pérez, F. and  Granger, B.E.
+               *IPython: A System for Interactive Scientific Computing*,
+               Computing in Science and Engineering, 9(3):21-29, May/June 2007
+
+.. [ccdproc] Crawford, S and Craig, M., https://github.com/ccdproc
+
+.. [ref_needed] *FILL ME IN*
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: This is still to-do

File: papers/matthew_craig/matthew_craig.rst (hunk 4 above)
Comment (taldcroft) [inline papers/matthew_craig/matthew_craig.rst]: I second this comment, I was definitely confused at this point.

**Comment by taldcroft** [comment]
@mwcraig - I'm also reviewing your paper.  It looks good and my only additional substantive comment is that I think you can strengthen the case for `reducer` over the other solutions (in particular AstroImageJ):

-  AstroImageJ is really complete and full-featured, but ultimately it is a complex black box that separates the student from the process.  There is no straightforward way to inspect or modify the processing algorithms.  By contrast `reducer` is something of a box, but with glass walls and an invitation to get in an tinker and understand.

- By using IPython notebook as the core interface, `reducer` introduces students to programming and Python in a non-intimidating way.  More advanced students can then naturally go from data reduction to Python data analysis, maybe even in the same notebook.

- For students that do intend to pursue postgraduate studies in science or engineering, having exposure to Python and programming will be advantageous.  Having experience with AstroImageJ will be of minimal value.

**Comment by eteq** [comment]
@mwcraig - I also did a read-through and don't have much to add beyond the existing comments - looks very nice as is!

My one comment is just a pedantic one: I notice you're not consistent about how you reference "affiliated packages".  I suggest just always saying they are "affiliated package(s)" or "Astropy affiliated package(s)" (you use "affiliate" with and without quotes in a few places).

File: papers/matthew_craig/matthew_craig.rst (hunk 8 above)
Comment (mwcraig) [inline papers/matthew_craig/matthew_craig.rst]: now done...

**Comment by mwcraig** [comment]
Several of the comments have been addressed:

+ @jeanconn -- I have room to add a few more screen shots side-by-side with code snippets to illustrate the basic structure. Will get that in tonight if I can.
+ @eteq -- affiliated/affiliate cleaned up, I think.
+ @taldcroft -- I will incorporate your suggestions tonight.

Thanks to everyone for the feedback!

**Comment by mwcraig** [comment]
Everything is now addressed except @jeanconn's concern about running out of illustrations by the time I get to the details of the widgets. I started working on that by fleshing out the widget structure section with a code example but it seems inelegant. Will sleep on it.

**Comment by jeanconn** [comment]
Feel free to ignore my previous comment suggesting more visual aids if it doesn't seem to work! 
I think this is good-to-go at this point.  

To some extent, I think I was looking for more in the widget description (either in visual aids or in text) to explain why this is a good and elegant solution, and you've done that by addressing @taldcroft's comments earlier in the paper.

File: papers/matthew_craig/matthew_craig.rst [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+:email: mcraig@mnstate.edu
+:institution: Department of Physics and Astronomy, Minnesota State University Moorhead
+
+--------------------------------------------------------------------------
+Widgets and Astropy: Accomplishing Productive Research with Undergraduates
+--------------------------------------------------------------------------
+
+.. class:: abstract
+
+    A set of IPython notebooks with a widget interface that are based on
```
Comment (jakevdp) [inline papers/matthew_craig/matthew_craig.rst]: The passive voice here makes the abstract hard for the reader to parse. Also, this is not fundamentally about IPython notebooks, but about a tool for astronomical research

I'd start with something like "This paper describes a tool for astronomical research implemented as a set of IPython notebooks with a widget interface..."

File: papers/matthew_craig/matthew_craig.rst [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+widget structure* presents its implementation. The widget classes in
+``reducer`` are potentially useful in other applications.
+
+.. [#] http://www.astropy.org/affiliated/
+
+Background: Image analysis in optical stellar astronomy
+-------------------------------------------------------
+
+While a detailed description of astronomical data analysis is beyond the scope
+of this talk, some appreciation of the steps involved is useful for
```
Comment (jakevdp) [inline papers/matthew_craig/matthew_craig.rst]: Beyond the scope of this paper?

File: papers/matthew_craig/matthew_craig.rst [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+It is the combination of high-quality python packages for both the back-end
+and front-end that made development of ``reducer`` relatively straightforward.
+
+A notebook-based solution offers a couple of other advantages over even the
+strongest of the GUI tools discussed in the previous section. The first is
+that exposure to programming broadly is useful to both the few students who
+become professional astronomers and the ones who do not. Though no programming
+is required to use ``reducer``, there is code in several of the notebook
+cells. It represents something intermediate between a fully GUI application
+and script-only interface. Another is that exposure to Python nprogramming is
```
Comment (jakevdp) [inline papers/matthew_craig/matthew_craig.rst]: typo: nprogramming -> programming

**Comment by jakevdp** [comment]
New version looks great! I have no substantive comments aside from the few minor typos pointed out above.

**Comment by mwcraig** [comment]
@jakevdp -- thanks for the feedback, all has been incorporated.

@jeanconn -- I added a code example with screenshot of the output to the last section. Let me know what you think.

all: no more changes are planned except in response to comments. Thanks again for all of your feedback!

**Comment by taldcroft** [comment]
Looks good to me.

**Comment by jeanconn** [comment]
I'd like to look over the PDF again but I'm just seeing empty brackets on the build page (and they persist after trying the refresh-build process) 

build

 @katyhuff @jaberg ?

**Comment by katyhuff** [comment]
@jeanconn - yeah, I'm seeing that too. It's weird. Let me restart the server.

**Comment by mwcraig** [comment]
There may be a server problem too, but I forgot to include the new figure in my last commit.

**Comment by katyhuff** [comment]
Hah, simultaneous fixes leave the masses still wondering: which one fixed the build?
Server back up, pdf available.

File: papers/matthew_craig/matthew_craig.rst [hunk 12]
```diff
@@ -0,0 +1,521 @@
...
+----------------------------
+
+
+At the base of the reducer widget structure is an extension of a container
+widget from IPython. This class, ``ToggleContainerWidget``, adds a toggle to
+control display the contents of the container, and a list of child widgets
+displayed in the container. [#]_  Since a ``ToggleContainerWidget`` can have
+another ``ToggleContainerWidget`` as a child, this immediately provides an
+interface for presenting a user with a nested list of options. Fig. `light-
+settings` has a thin red border drawn around each element that is a subclass
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: I think you need the ':ref:' bit here to have light-settings resolve to figure 4.

File: papers/matthew_craig/matthew_craig.rst [hunk 13]
```diff
@@ -0,0 +1,521 @@
...
+    can be displayed.
+    :label:`image-display-example`
+
+``reducer`` widget structure
+----------------------------
+
+
+At the base of the reducer widget structure is an extension of a container
+widget from IPython. This class, ``ToggleContainerWidget``, adds a toggle to
+control display the contents of the container, and a list of child widgets
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: Probably should be "control display *of* the contents".

File: papers/matthew_craig/matthew_craig.rst [hunk 14]
```diff
@@ -0,0 +1,521 @@
...
+            c_box = CheckboxWidget(description="Don't check me")
+
+            # These children are contained in the
+            # MyControl widget
+            self.add_child(b_box)
+            self.add_child(c_box)
+
+
+An ``is_sane`` property that can be overridden by subclasses to indicate that
+the settings in the widget are sensible. This can provide some minimal
```
Comment (jeanconn) [inline papers/matthew_craig/matthew_craig.rst]: This sentence ended up as a sentence fragment.

**Comment by jeanconn** [comment]
Looks great!  I think the additional examples, concepts, and figure in the widget nuts-and-bolts section will help somebody else looking to develop along these lines.

I noted those additional tiny typos when doing closer reading in the widget structure section around the new figure.

And a last tiny question: should there be a link to https://github.com/mwcraig/reducer somewhere in the paper?

**Comment by mwcraig** [comment]
@jeanconn @kbarbary @jakevdp @eteq @taldcroft -- all revisions have been made. Thank you again for your helpful suggestions!

@katyhuff @jaberg -- I believe I have address all of the reviewers comments.

**Comment by katyhuff** [comment]
@mwcraig : It certainly looks like you've satisfied all the reviewer comments as far as I can tell. I'll change this tag to ready. If any reviewers (@jeanconn @kbarbary @jakevdp @eteq @taldcroft ) notice any remaining unaddressed recommendations essential for acceptance, please pipe up in the next couple of days. 

To all involved in this particular review: Thanks for being in efficient contact with one another as changes were suggested and made. It clearly made the process go very smoothly for all. Nicely done all around!

**Comment by jakevdp** [comment]
@katyhuff @jaberg -- This looks good from my end! Thanks!

**Pull request merged**